File: Python/best_time_buy_sell_stock.py
class Solution(object):
    def __init__(self):
        self.name = 'S->13'

    # Time complexity: O(n)
    # Space complexity: O(1)
    def maxProfit(self, prices):
        start = prices[0]
        maxPro = 0
        for i in range(1, len(prices)):
            if prices[i] < prices[i - 1]:
                maxPro = max(maxPro, prices[i - 1] - start)
                start = min(start, prices[i])

        return max(maxPro, prices[-1] - start)

prices = [2,3,5,8,7,1,5,3,6,4]
print(Solution().maxProfit(prices))

File: Algorithm/DP/best_time_buy_sell_stock.cpp
class Solution {
public:
  // Time complexity: O(n)
  // Space complexity: O(1)
  int maxProfit(vector<int>& prices) {
    int i = 0;
    int n = prices.size();
    int profit = 0;
    while (i < n) {
      while (i < n-1 && prices[i] >= prices[i+1]) i++;
      int buy = prices[i];
      while (i < n-1 && prices[i] < prices[i+1]) i++;
      int sell = prices[i];
      profit += (sell - buy);
      if(i == n-1) break;
    }
    return profit;
  }
};
